### ISSUE TITLE
"please activate your account"

### ISSUE THREAD

**Issue by gerberanub** [Misioslav/expressvpn#31]
root@3da5b1035cc0:/# expressvpn connect
Connecting to Smart Location...
Connecting to Japan - Shibuya...
Unable to connect to Japan - Shibuya

   - Check your internet connection or try a different location.
   - Internet traffic is blocked to protect your privacy.
   - To retry the connection, type 'expressvpn connect'
   - To disconnect from the VPN and unblock internet traffic, type 'expressvpn disconnect'
root@3da5b1035cc0:/# expressvpn diagnostics
Client Version: 0.0.1default
Client Shared Version: v23.21.0
OS Name: linux
OS Version: Debian GNU/Linux 11
Internal diagnostics data:
D01: 448, OK
D02: 447, OK
D03: 438
D04: 438, OK
D05: 438, OK
D06: 438
D07: 438, OK
D08: 438, OK
D09: 86707912
D10: 438, OK
D11: 438, OK
D12: N, N
D13: www.2spb2qq8pnqmv.net/api
D14: Japan - Shibuya
D15: auto
D17: 438, OK
D18: 0
Network Lock:         enabled, allow access to local devices
DNS leak protection:  enabled
IPv6 leak protection: enabled
Selected location:    Japan - Shibuya (Smart Location)

==============================================

Connecting to Japan - Shibuya, ip: 24.123.193.169, protocol: lightway_udp

Client Version: 0.0.1default
Client Shared Version: v23.21.0
OS Name: linux
OS Version: Debian GNU/Linux 11
Internal diagnostics data:
D01: 432, OK
D02: 431, OK
D03: 422
D04: 422, OK
D05: 422, OK
D06: 422
D07: 422, OK
D08: 422, OK
D09: 86707896
D10: 422, OK
D11: 422, OK
D12: N, N
D13: www.2spb2qq8pnqmv.net/api
D14: Japan - Shibuya
D15: auto
D17: 422, OK
D18: 0

can't connect please help me

**Comment by Misioslav** [comment]
Hi, what is the exact issue you are facing? I understand you can't connect but it might be caused by a couple of things. The title would suggest you need to have your account activated for expressvpn. Did you already purchase your license key? If you did, did you set it up as env var for your docker configuration?
What machine are you using? Is it Synology, QNAP or maybe a custom build?

**Comment by baszondag** [comment]
Hi, 
im experiencing the same issue. 
i cannot connect to to smart location and im using a Asustor NAS 

The diagnostics give the same output, but still cannot connect the VPN

**Comment by Misioslav** [comment]
Please, post your configuration of your docker or compose file

**Comment by baszondag** [comment]
here is my compose

  expressvpn:
    container_name: expressvpn
    image: misioslav/expressvpn:latest-bookworm  
    environment:
      - ACTIVATION_CODE=${EXPRESSVPN-CODE}
      - SERVER=${EXPRESSVPN_LOCATION}
      - PREFERRED_PROTOCOL=${EXPRESSVPN_PROTOCOL}
      - LIGHTWAY_CIPHER=${EXPRESSVPN_CIPHER}
    cap_add:
      - NET_ADMIN
    devices: 
      - /dev/net/tun
    stdin_open: true
    tty: true
    command: /bin/bash
    privileged: true
    restart: unless-stopped
    networks:
      - default

EXPRESSVPN_LOCATION=Austria
EXPRESSVPN_PROTOCOL=lightway_udp
EXPRESSVPN_CIPHER=chacha20

**Comment by Misioslav** [comment]
For my docker the environments you are using are incorrect and should be changed:
`ACTIVATION_CODE` should be `CODE`
`PREFERRED_PROTOCOL` should be `PROTOCOL`
`LIGHTWAY_CIPHER` should be `CIPHER`

but the last two are optional so they won't bear much impact but the CODE env will.  Please, change the name of env for code and try again.

**Comment by Webreaper** [comment]
I'm also seeing this. I checked my activation code, and it's definitely correct. Docker-compose: 
```
 expressvpn:
        image: misioslav/expressvpn:latest
        container_name: expressvpn
        environment:
            - CODE={super-sekret-code}
        cap_add:
            - NET_ADMIN
        devices:
            - "/dev/net/tun:/dev/net/tun"
        stdin_open: true
        privileged: true
        tty: true
        ports:
            - 8090:8090 
            - 6881:6881
            - 6881:6881/udp
            - 8080:8080
            - 9117:9117
            - 9696:9696
        command: /bin/bash
        restart: unless-stopped
```
Any ideas?

**Comment by Misioslav** [comment]
> I'm also seeing this. I checked my activation code, and it's definitely correct. Docker-compose:
> 
> ```
>  expressvpn:
>         image: misioslav/expressvpn:latest
>         container_name: expressvpn
>         environment:
>             - CODE={super-sekret-code}
>         cap_add:
>             - NET_ADMIN
>         devices:
>             - "/dev/net/tun:/dev/net/tun"
>         stdin_open: true
>         privileged: true
>         tty: true
>         ports:
>             - 8090:8090 
>             - 6881:6881
>             - 6881:6881/udp
>             - 8080:8080
>             - 9117:9117
>             - 9696:9696
>         command: /bin/bash
>         restart: unless-stopped
> ```
> 
> Logs:
> 
> <img alt="Screenshot 2024-01-19 at 10 48 25" width="1242" src="https://private-user-images.githubusercontent.com/4901102/298039294-59a0f68c-ef43-456b-946d-d4458b10ce82.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU2NjMzNjIsIm5iZiI6MTcwNTY2MzA2MiwicGF0aCI6Ii80OTAxMTAyLzI5ODAzOTI5NC01OWEwZjY4Yy1lZjQzLTQ1NmItOTQ2ZC1kNDQ1OGIxMGNlODIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDExOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMTlUMTExNzQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDZiZjVhYTNlNzc3ZmE5NWYxYmZlZGM4NDhkNjgwN2I0Y2VmN2YwOTA1N2Y5MjZmYWJhMzE3NzRkNTY3NzYxNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.YhCSu0Mw9EDmfxEIdhHlZCoHPqr1c190FEpo9Ewts70">
> Any ideas? If it makes a difference, I'm running on a Synology DS1520.

Occasionally, it also happens to me, but it usually works after a restart. Honestly, I never had time to try debugging, especially since it happened so rarely that it was hard to reproduce.
I understand that in your case your CODE is in the correct format nor include any whitespace or anything. Also, a container restart doesn't do the trick, right?

**Comment by Webreaper** [comment]
So, interesting info. Restarting doesn't work at all, whatever I do. Tried many times. 

However, if I do this:

```
sudo docker exec -it expressvpn bash
root@3baf5974ccf8:/# cd expressvpn/
root@3baf5974ccf8:/expressvpn# bash start.sh 
```
It seems to start working - I see the following output:

```
umount: /etc/resolv.conf: not mounted.
Preferred Protocol is lightway_udp.
LightwayCipher is chacha20.
Send Diagnostics is disabled.
Threat Manager is enabled.
Kernel Version is lower than minimum version of required kernel (4.9), network_lock will be disabled.
Network Lock Mode is off.
Auto Connect is enabled.
ExpressVPN for Linux v3.61.0.12 is now available.

To update ExpressVPN, run sudo apt update && sudo apt install --only-upgrade expressvpn
If you encounter an issue, download the app manually at https://www.vlycgtx.com/latest#linux

If your VPN connection unexpectedly drops, internet traffic will now be
blocked by Network Lock to protect your privacy.

Connecting to Smart Location...
Connecting to UK - East London...	15.0%
```
(it doesn't seem to get past 15%, but unsure if that's expected). 

So i'm wondering if it's some sort of race condition or other path/config thing?

**Comment by Webreaper** [comment]
Restarting the container results in it attempting to connect. So could be a race condition first time it launches. However, It fails to connect..... trying a different location and protocol. Does your container support the `wireguard` protocol option?

**Comment by Misioslav** [comment]
> Restarting the container results in it attempting to connect. So could be a race condition first time it launches. However, It fails to connect..... trying a different location and protocol. Does your container support the `wireguard` protocol option?

For connection issues, usually those were connected to tun settings, please, check the script from there https://github.com/Misioslav/expressvpn/issues/11#issuecomment-1003757934 and see if that works for you.

As for the activation issue, it might be the race condition. I would have to try to look more closely into that but unfortunately, lack of time limits me right now.

**Comment by Webreaper** [comment]
Not getting the tun issue per that link. It just doesn't connect. See below:

<img width="1143" alt="Screenshot 2024-01-19 at 11 53 44" src="https://github.com/Misioslav/expressvpn/assets/4901102/89c65ff0-7de0-408c-bdee-91b4208091ff">

Trying a different protocol and location.

**Comment by Webreaper** [comment]
Okay, you may be right. Running that tun fix script seems to have helped, I think. 

BTW, your container uses expressvpn v3.6, whereas there's a new version 3.61 which seems better. Is it worth doing a new build? And perhaps adding an upgrade script so when the container starts it upgrades to the latest client?

**Comment by Misioslav** [comment]
> Okay, you may be right. Running that tun fix script seems to have helped, I think.
> 
> BTW, your container uses expressvpn v3.6, whereas there's a new version 3.61 which seems better. Is it worth doing a new build? And perhaps adding an upgrade script so when the container starts it upgrades to the latest client?

Ive already pushed newest build so it should be available.
As for the upgrade script I was thinking about to add it but again it goes into having time territory, so I will for now add it to to do list.
Maybe in the upcoming week or two I will be able to work on it.

**Comment by Webreaper** [comment]
Cool, thanks. I'm pulling latest and it seems to be there.

It seems to be working for me now, other than the fact that I have to restart the container after the initial pull/docker-compose....

It's just annoying that Synology's kernel is old so the network_lock doesn't work. :(

**Comment by Misioslav** [comment]
Some time ago I found a DNS leak test (sh script) that I use to check if all is fine with my container and other containers that are using it, and if not then I do additional actions. Also, if you use just a couple of containers to with it not the whole machine I recommend also copying resolv.conf from expressvpn to the containers that are connected to it to additionally secure DNS leaking :)

```
#!/usr/bin/env bash
#usage:   ./dnsleaktest.sh [-i interface_ip|interface_name]
#example: ./dnsleaktest.sh -i eth1
#         ./dnsleaktest.sh -i 10.0.0.2

RED='\033[0;31m'
BOLD='\033[1m'
NC='\033[0m'
api_domain='bash.ws'
error_code=1

getopts "i:" opt
interface=$OPTARG

function echo_bold {
    echo -e "${BOLD}${1}${NC}"
}

if [ -z "$interface" ]; then
    curl_interface=""
    ping_interface=""
else
    curl_interface="--interface ${interface}"
    ping_interface="-I ${interface}"
    echo_bold "Interface: ${interface}"
    echo ""
fi

function increment_error_code {
    error_code=$((error_code + 1))
}

function echo_error {
    (>&2 echo -e "${RED}${1}${NC}")
}

function require_command {
    command -v $1 > /dev/null
    if [ $? -ne 0 ]; then
        echo_error "Please, install \"$1\""
        exit $error_code
    fi
    increment_error_code
}

function check_internet_connection {
    curl --silent --head ${curl_interface} --request GET "https://${api_domain}" | grep "200 OK" > /dev/null
    if [ $? -ne 0 ]; then
        echo_error "No internet connection."
        exit $error_code
    fi
    increment_error_code
}

require_command curl
require_command ping
check_internet_connection

if command -v jq &> /dev/null; then
    jq_exists=1
else
    jq_exists=0
fi

if hash shuf 2>/dev/null; then
    id=$(shuf -i 1000000-9999999 -n 1)
else
    id=$(jot -w %i -r 1 1000000 9999999)
fi

for i in $(seq 1 10); do
    ping -c 1 ${ping_interface} "${i}.${id}.${api_domain}" > /dev/null 2>&1
done

function print_servers {

    if (( $jq_exists )); then

        echo ${result_json} | \
            jq  --monochrome-output \
            --raw-output \
            ".[] | select(.type == \"${1}\") | \"\(.ip)\(if .country_name != \"\" and  .country_name != false then \" [\(.country_name)\(if .asn != \"\" and .asn != false then \" \(.asn)\" else \"\" end)]\" else \"\" end)\""

    else

        while IFS= read -r line; do
            if [[ "$line" != *${1} ]]; then
                continue
            fi

            ip=$(echo $line | cut -d'|' -f 1)
            code=$(echo $line | cut -d'|' -f 2)
            country=$(echo $line | cut -d'|' -f 3)
            asn=$(echo $line | cut -d'|' -f 4)

            if [ -z "${ip// }" ]; then
                 continue
            fi

            if [ -z "${country// }" ]; then
                 echo "$ip"
            else
                 if [ -z "${asn// }" ]; then
                     echo "$ip [$country]"
                 else
                     echo "$ip [$country, $asn]"
                 fi
            fi
        done <<< "$result_txt"

    fi
}


if (( $jq_exists )); then
    result_json=$(curl ${curl_interface} --silent "https://${api_domain}/dnsleak/test/${id}?json")
else
    result_txt=$(curl ${curl_interface} --silent "https://${api_domain}/dnsleak/test/${id}?txt")
fi

dns_count=$(print_servers "dns" | wc -l)

echo_bold "Your IP:"
print_servers "ip"

echo ""
if [ ${dns_count} -eq "0" ];then
    echo_bold "No DNS servers found"
else
    if [ ${dns_count} -eq "1" ];then
        echo_bold "You use ${dns_count} DNS server:"
    else
        echo_bold "You use ${dns_count} DNS servers:"
    fi
    print_servers "dns"
fi

echo ""
echo_bold "Conclusion:"
print_servers "conclusion"

exit 0
```

**Comment by burnbrigther** [comment]
Hi there - I can start a new thread on this - but I too am running in to the activation issue.  I can go within the docker container and fix it by reconnecting, but it doesn't persist reboots.  What is the "tun" fix described above?  It is either gone or blocked in github.  Thanks for your efforts on this.  I'm hoping the performance is better than gluetun and expressvpn on openvpn.

**Comment by Misioslav** [comment]
> Hi there - I can start a new thread on this - but I too am running in to the activation issue. I can go within the docker container and fix it by reconnecting, but it doesn't persist reboots. What is the "tun" fix described above? It is either gone or blocked in github. Thanks for your efforts on this. I'm hoping the performance is better than gluetun and expressvpn on openvpn.

For tun I use this

https://github.com/Misioslav/expressvpn/issues/11#issuecomment-1003757934

as for the activation it is at least for me persistent through the restarts but at first run it may fail. I still need to find some time to try to address it somehow

**Comment by burnbrigther** [comment]
I am wondering if it can be addressed by doing an "expressvpn connect" as a part of the initialization command sequence.  That's a pretty bad hack, but it may work until the real problem can be found.

**Comment by burnbrigther** [comment]
Ok, I can see this is an issue with the ENTRY script, which is trying to do just that.  There is more information on the [polkaned/expressvpn] issues side.

**Comment by Misioslav** [comment]
> Ok, I can see this is an issue with the ENTRY script, which is trying to do just that. There is more information on the [polkaned/expressvpn] issues side.

Ive pushed some changes to activate steps. Please, check if that will work for you.

**Comment by burnbrigther** [comment]
Hey, nice!  That works.  Thank you!  Perfect.

**Comment by Misioslav** [comment]
Thank you for your kind words. 
I’m glad it worked 😊

**Comment by Webreaper** [comment]
Not sure what you changed, but it's worse for me. I can't seem to get connected - even if I restart the container to work around the known race condition. :(

**Comment by burnbrigther** [comment]
This is what is working for me.  And the only way it worked for me.  There must be something else going on.  Activation was continually being asked for my connection.

**Comment by Webreaper** [comment]
Let me try that.

**Comment by Misioslav** [comment]
But then it’s a different issue. The activation went correct you cannot connect. Usually for me it was a tun issue I had to fix with this script

https://github.com/Misioslav/expressvpn/issues/11#issuecomment-1003757934

**Comment by Webreaper** [comment]
Oh, yes, maybe I haven't run that recently. Let me check

**Comment by Webreaper** [comment]
Sorry, you're absolutely right. My bad. Have put that in my docker-compose script so it never catches me out again. 

BTW, I tried AUTO_UPDATE=on, but it seemed to just infinitely loop. But that's a different issue.

**Comment by Misioslav** [comment]
> Sorry, you're absolutely right. My bad. Have put that in my docker-compose script so it never catches me out again.

Great! Im happy it works. I also have it my startup scripts that I run at least once a week to also update all containers.

> BTW, I tried AUTO_UPDATE=on, but it seemed to just infinitely loop. But that's a different issue.

In such case if you may, please, create a new issue and provide me with logs, I will take a look at it. I had some "looping" issue with it for a moment but some changes Ive done recently seemed to stop that behavior and now, at least on my machine, it works without a hiccup.
